### PR TITLE
Estimate cleaning threshold with CatB pedestals if availables

### PR DESCRIPTION
**Pull request by FrancaCassol**

### CODE REVIEW

File: lstchain/io/io.py [hunk 1]
```diff
@@ -77,6 +77,9 @@
 dl1_params_tel_mon_ped_key = "/dl1/event/telescope/monitoring/pedestal"
 dl1_params_tel_mon_cal_key = "/dl1/event/telescope/monitoring/calibration"
 dl1_params_tel_mon_flat_key = "/dl1/event/telescope/monitoring/flatfield"
+dl1_params_tel_mon_catB_ped_key = "/dl1/event/telescope/monitoring/CatB/pedestal"
```
Comment (maxnoe) [inline lstchain/io/io.py:80]: This key is not correct (I assume it is the one that the calibration tool is writing, in that case, I missed it during review, sorry).

It should be `/dl1/monitoring/telescope/...` (not have event!)

Comment (FrancaCassol) [reply lstchain/io/io.py:80]: I see, I will update it

File: lstchain/scripts/lstchain_dl1ab.py [hunk 2]
```diff
@@ -44,6 +44,9 @@
     dl1_params_lstcam_key,
     global_metadata,
     write_metadata,
+    dl1_params_mon_tel_catB_ped_key,
+    dl1_params_mon_tel_catB_ff_key,
```
Comment (aaguasca) [inline lstchain/scripts/lstchain_dl1ab.py:48]: This should be `dl1_params_mon_tel_catB_flat_key`, right?

Comment (FrancaCassol) [reply lstchain/scripts/lstchain_dl1ab.py]: yes, thanks for spotting it

File: lstchain/scripts/lstchain_dl1ab.py [hunk 3]
```diff
@@ -325,6 +331,14 @@ def main():
                     image_table['image'][ii] = image
                     image_table['peak_time'][ii] = peak_time
 
+                    # use CatB pedestals to estimate the picture threshold
+                    if args.pedestal_cleaning:
+                        ped_charge_mean_pe = catB_pedestal_mean[calib_idx][selected_gain, pixel_index]
+                        ped_charge_std_pe = catB_pedestal_std[calib_idx][selected_gain, pixel_index]
```
Comment (aaguasca) [inline lstchain/scripts/lstchain_dl1ab.py:337]: The conversion to p.e. is not done before

```
ped_charge_mean_pe = catB_pedestal_mean[calib_idx][selected_gain, pixel_index] * dc_to_pe
ped_charge_std_pe = catB_pedestal_std[calib_idx][selected_gain, pixel_index] * dc_to_pe
```

File: lstchain/scripts/lstchain_dl1ab.py [hunk 4]
```diff
@@ -325,6 +331,14 @@ def main():
...
+                        threshold_clean_pe = ped_charge_mean_pe + sigma * ped_charge_std_pe
+                        threshold_clean_pe[unusable_pixels] = pic_th
+                        picture_th = np.clip(pedestal_thresh, pic_th, None)
```
Comment (aaguasca) [inline lstchain/scripts/lstchain_dl1ab.py:340]: I am not sure I am following this part. I expected something like:

```python
threshold_clean_pe[unusable_pixels] = max(threshold_clean_pe)
picture_th = np.clip(threshold_clean_pe, pic_th, None)
```

As it is done in `get_threshold_from_dl1_file`.

File: lstchain/scripts/lstchain_dl1ab.py (hunk 3 above)
Comment (FrancaCassol) [reply lstchain/scripts/lstchain_dl1ab.py:337]: the pedestals are extracted from calibrated waveforms, so this is not necessary

File: lstchain/scripts/lstchain_dl1ab.py (hunk 4 above)
Comment (FrancaCassol) [reply lstchain/scripts/lstchain_dl1ab.py:340]: the unusable pixels are set to zero in the Cat-B process, so in principle the pic_th is enough to cut them...

**Comment by FrancaCassol** [comment]
Hi @maxnoe,
If nothing against could we merge this request?

File: lstchain/io/io.py [hunk 5]
```diff
@@ -77,6 +77,9 @@
 dl1_params_tel_mon_ped_key = "/dl1/event/telescope/monitoring/pedestal"
 dl1_params_tel_mon_cal_key = "/dl1/event/telescope/monitoring/calibration"
 dl1_params_tel_mon_flat_key = "/dl1/event/telescope/monitoring/flatfield"
+dl1_params_mon_tel_catB_ped_key = "/dl1/monitoring/telescope/CatB/pedestal"
+dl1_params_mon_tel_catB_cal_key = "/dl1/monitoring/telescope/CatB/calibration"
+dl1_params_mon_tel_catB_flat_key = "/dl1/monitoring/telescope/CatB/flatfield"
```
Comment (morcuended) [inline lstchain/io/io.py:82]: these variables are being called differently in #1147 and #1174, without `_params`

File: lstchain/scripts/lstchain_dl1ab.py [hunk 6]
```diff
@@ -44,6 +44,9 @@
     dl1_params_lstcam_key,
     global_metadata,
     write_metadata,
+    dl1_params_mon_tel_catB_ped_key,
```
Comment (morcuended) [inline lstchain/scripts/lstchain_dl1ab.py:47]: see comment above

File: lstchain/scripts/lstchain_dl1ab.py [hunk 7]
```diff
@@ -431,9 +448,9 @@ def main():
 
             # write a cat-B calibrations in DL1b
             if catB_calib:
-                write_table(catB_calib, outfile, "/dl1/event/telescope/monitoring/catB/calibration")
-                write_table(catB_pedestal, outfile, "/dl1/event/telescope/monitoring/catB/pedestal")
-                write_table(catB_flatfield, outfile, "/dl1/event/telescope/monitoring/catB/flatfield")
+                write_table(catB_calib, outfile, dl1_params_mon_tel_catB_cal_key)
```
Comment (morcuended) [inline lstchain/scripts/lstchain_dl1ab.py:451]: naming of the catB keys

File: lstchain/scripts/lstchain_dl1ab.py [hunk 8]
```diff
@@ -325,6 +336,12 @@ def main():
                     image_table['image'][ii] = image
                     image_table['peak_time'][ii] = peak_time
 
+                    # use CatB pedestals to estimate the picture threshold
+                    if args.pedestal_cleaning:
+                        threshold_clean_pe = catB_threshold_clean_pe[calib_idx][selected_gain, pixel_index]
+                        threshold_clean_pe[unusable_pixels] = pic_th
```
Comment (morcuended) [inline lstchain/scripts/lstchain_dl1ab.py:342]: I suggest adding a comment that for unusable pixels the picture threshold is the one set in the config file

**Review by morcuended — COMMENTED** [review]
Thanks @FrancaCassol. I left a few comments. Is it possible to implement a test for this?

**Comment by FrancaCassol** [comment]
Hi @morcuended, for implementing a test, we need first to merge PR [https://github.com/cta-observatory/cta-lstchain/pull/1147](https://github.com/cta-observatory/cta-lstchain/pull/1147), in order to easily produce a catB calibration file.

**Review by morcuended — DISMISSED** [review]

**Comment by rlopezcoto** [comment]
@FrancaCassol I resolved the conflicts, can you check if everything is alright?

**Comment by FrancaCassol** [comment]
Hi @rlopezcoto,
perhaps in order to be coherent, we should  keep the "C" of CatB, uppercase or lowercase in the key name and  in the key path:

dl1_mon_tel_catB_ped_key = "/dl1/monitoring/telescope/catB/pedestal"
dl1_mon_tel_catB_cal_key = "/dl1/monitoring/telescope/catB/calibration"
dl1_mon_tel_catB_flat_key = "/dl1/monitoring/telescope/catB/flatfield"

or 

dl1_mon_tel_CatB_ped_key = "/dl1/monitoring/telescope/CatB/pedestal"
dl1_mon_tel_CatB_cal_key = "/dl1/monitoring/telescope/CatB/calibration"
dl1_mon_tel_CatB_flat_key = "/dl1/monitoring/telescope/CatB/flatfield"

**Comment by rlopezcoto** [comment]
> in the hdf5 files merging script is still written as `dl1_mon_tel_CatB_ped_key`

corrected now

File: lstchain/scripts/lstchain_dl1ab.py [hunk 9]
```diff
@@ -184,11 +185,18 @@ def main():
...
+        log.info(f"Fraction of Cat_A pixel cleaning thresholds above Cat_A picture thr.:"
                  f"{np.sum(pedestal_thresh > pic_th) / len(pedestal_thresh):.3f}")
         picture_th = np.clip(pedestal_thresh, pic_th, None)
         log.info(f"Tailcut clean with pedestal threshold config used:"
                  f"{config['tailcuts_clean_with_pedestal_threshold']}")
+        
+        if args.catB_calibration_file is not None:
+            catB_pedestal_mean = np.array(catB_pedestal["charge_mean"])
+            catB_pedestal_std= np.array(catB_pedestal["charge_std"])
+            catB_threshold_clean_pe = catB_pedestal_mean + sigma * catB_pedestal_std
```
Comment (morcuended) [inline lstchain/scripts/lstchain_dl1ab.py:197]: Maybe these lines could be somehow integrated inside `get_threshold_from_dl1_file`, which I think does something similar.

**Review by morcuended — APPROVED** [review]
I left a minor comment that can be ignored. Looks good to me

**Pull request merged**